### PR TITLE
bpf documentation updates, part 5

### PR DESCRIPTION
**Pull request by borkmann**
This concludes the initial version of the BPF / XDP ref guide writeup.

Fixes: #713

### CODE REVIEW

**Comment by borkmann** [comment]
test-docs-please

**Comment by borkmann** [comment]
test-docs-please

**Comment by borkmann** [comment]
test-docs-please

**Comment by borkmann** [comment]
test-docs-please

File: Documentation/bpf.rst [hunk 1]
```diff
@@ -3354,7 +3354,354 @@ the toolchain section under the respective tools.
...
+  the network packet representation of the kernel's networking stack and
+  therefore used in many places. As such it has all the relevant ``sk_buff``
+  metadata available the BPF program can read and write into, whereas in
+  ``xdp_buff`` case it's limited due to the early stage processing where the
+  cost of allocating a ``sk_buff`` is avoided altogether.
+
+  Out of the tc BPF hooks, the BPF program can for instance read or write
+  the skb's ``mark``, ``pkt_type``, ``protocol``, ``priority``, ``queue_mapping``,
+  ``napi_id``, ``cb[]`` array, ``hash``, ``tc_classid`` or ``tc_index``, vlan
+  metadata, the XDP transferred custom metadata and various other information.
```
Comment (joestringer) [inline Documentation/bpf.rst]: Could potentially link to the file which defines this stuff.

File: Documentation/bpf.rst [hunk 2]
```diff
@@ -3354,7 +3354,354 @@ the toolchain section under the respective tools.
...
+
+  The ``sk_buff`` is of a completely different nature than ``xdp_buff`` where
+  both come with advantages and disadvantages. For example, the ``sk_buff``
```
Comment (joestringer) [inline Documentation/bpf.rst]: s/sk_buff/xdp_buff/

**Comment by borkmann** [comment]
test-docs-please

**Comment by borkmann** [comment]
test-docs-please

File: Documentation/bpf.rst [hunk 3]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+kernel's tc (traffic control) layer in the networking data path. On a high-level
+there are three major differences when comparing BPF XDP programs to tc BPF
+ones:
+
+* The BPF input context is a ``sk_buff`` not a ``xdp_buff``. The former is
+  the network packet representation of the kernel's networking stack and
+  therefore used in many places. As such it has all the relevant ``sk_buff``
+  metadata available the BPF program can read and write into, whereas in
+  ``xdp_buff`` case it's limited due to the early stage processing where the
+  cost of allocating a ``sk_buff`` is avoided altogether.
```
Comment (joestringer) [inline Documentation/bpf.rst]: I'm not sure why it's important to say `therefore used in many places` above. If that part of the sentence was meant to lead to something perhaps you can add that part, otherwise I'd suggest dropping that part of that sentence.

I wonder if some wording like this might be clearer?

```
The BPF input context is a ``sk_buff`` not a ``xdp_buff``. When the kernel's networking stack receives a packet, after XDP processing, it allocates a buffer and parses the packet to store metadata about the packet. This representation is known as the ``sk_buff``. This structure is then exposed in the BPF input context so that BPF programs can use the metadata that the stack extracts from the packet. This can be useful, but comes with an associated costs of the stack performing this allocation and metadata extraction, and handling the packet until it hits the tc hook. By definition, the ``xdp_buff`` doesn't have access to this metadata because the XDP hook is called before this work is done. This is a significant contributor to the performance difference between the XDP and tc hooks.
```

File: Documentation/bpf.rst [hunk 4]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+  the skb's ``mark``, ``pkt_type``, ``protocol``, ``priority``, ``queue_mapping``,
+  ``napi_id``, ``cb[]`` array, ``hash``, ``tc_classid`` or ``tc_index``, vlan
+  metadata, the XDP transferred custom metadata and various other information.
+
+  The ``sk_buff`` is of a completely different nature than ``xdp_buff`` where
+  both come with advantages and disadvantages. For example, the ``sk_buff``
+  case has the advantage that it is rather straight forward to mangle its
+  associated metadata, however the kernel's ``sk_buff`` contains a lot of
+  protocol specific information which makes it difficult to simply switch
+  protocols by just rewriting the packet data. Thus, additional conversion
```
Comment (joestringer) [inline Documentation/bpf.rst]: I don't think that this quite clearly explains the underlying reason why the fact that the ``sk_buff`` containing protocol information makes it difficult to switch protocols. That reason being presumably that when the BPF program completes, the stack will continue processing the packet based on the metadata rather than just the packet contents.

File: Documentation/bpf.rst [hunk 5]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+  is required from BPF helper functions taking care that ``sk_buff`` internals
+  are properly converted as well. The ``xdp_buff`` case however does not
+  face such issues since it comes at such an early stage where the kernel
+  has not even allocated an ``sk_buff`` yet, thus packet rewrites of any
+  kind can be realized trivially. However, the ``xdp_buff`` case has the
+  disadvantage that ``sk_buff`` metadata is not available for mangling
+  at this time. The latter is overcome by passing custom metadata from
+  XDP BPF to tc BPF, though. Both program types can therefore be operated
+  complementary to overcome each limitation when required from the use case.
```
Comment (joestringer) [inline Documentation/bpf.rst]: I think that for `complementary` to become an adverb, it should be something like `complementarily`, but my web browser tells me that this isn't actually a word. Perhaps,

```
In this way, the limitations of each program type can be overcome by operating complementary programs of both types as the use case requires.
```

Not sure whether the `as the use case requires` needs to be there though.

File: Documentation/bpf.rst [hunk 6]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+
+..
+
+* The tc BPF programs do not require any driver changes since they are run
```
Comment (joestringer) [inline Documentation/bpf.rst]: `s/programs do/hook does/` ? (Would also need to remove the `hook` below, too.)

File: Documentation/bpf.rst [hunk 7]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+  at hook points in generic layers in the networking stack. Therefore, they
+  can be attached to any type of networking device.
+
+  While this provides flexibility, it also trades off performance compared
+  to running at the native XDP layer. However, tc BPF programs still come
+  at the earliest point in the generic kernel's networking data path after
+  GRO has been run but **before** any protocol processing, traditional iptables
+  firewalling such as iptables PREROUTING or nftables ingress hooks or other
+  packet processing takes place. Likewise on egress, at the latest point
```
Comment (joestringer) [inline Documentation/bpf.rst]: `Likewise on egress,` *tc BPF programs execute* `at the latest point...`

File: Documentation/bpf.rst [hunk 8]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+* Compared to XDP, tc BPF programs can be triggered out of ingress and also
+  egress points in the networking data path as opposed to ingress only in case
+  of XDP.
+
+  The two hook points ``sch_handle_ingress()`` and ``sch_handle_egress()`` in
+  the kernel are triggered out of ``__netif_receive_skb_core()`` and
+  ``__dev_queue_xmit()``, respectively. The latter two are the main receive
+  and transmit functions in the data path that, taking XDP aside, are triggered
+  for every network packet going in or coming out of the node allowing for
+  full visibility for tc BPF programs at these hook points.
```
Comment (joestringer) [inline Documentation/bpf.rst:3411]: I would consider shifting the paragraphs of this point above the next point up, since the next point up refers to egress as well.

File: Documentation/bpf.rst [hunk 9]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+
+..
+
+When talking about tc BPF programs in this context, then those are BPF programs
+which are run from ``cls_bpf`` classifier. Albeit tc terminology, the name
```
Comment (joestringer) [inline Documentation/bpf.rst]: I found this a little hard to parse. Perhaps,

```
BPF programs run in the tc hook are run from the ``cls_bpf`` classifier. While the tc terminology describes the BPF attachment point as a "classifier", this is is a bit misleading...
```

File: Documentation/bpf.rst [hunk 10]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+"classifier" is a bit misleading, since it underrepresents what ``cls_bpf``
```
Comment (joestringer) [inline Documentation/bpf.rst]: `under-represents`?

File: Documentation/bpf.rst [hunk 11]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+is capable of, that is, a fully programmable packet processor being able not
```
Comment (joestringer) [inline Documentation/bpf.rst]: `capable of` . *That is to say,* `a fully programmable...`

File: Documentation/bpf.rst [hunk 12]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+only to read the ``skb`` metadata and packet data but to also arbitrarily
```
Comment (joestringer) [inline Documentation/bpf.rst]: `not only to  ...` , `but also to ...`, `and ...` 

(commas)

File: Documentation/bpf.rst [hunk 13]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+mangle both and terminate the tc processing with an action verdict. ``cls_bpf``
+can thus be regarded as a self-contained entity managing and executing tc
```
Comment (joestringer) [inline Documentation/bpf.rst]: `self-contained entity` *that manages and executes* `tc BPF programs`.

File: Documentation/bpf.rst [hunk 14]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+BPF programs.
+
+It can hold one or multiple tc BPF programs. In the case where Cilium deploys
```
Comment (joestringer) [inline Documentation/bpf.rst]: s/It/``cls_bpf``/

s/multiple/more/

File: Documentation/bpf.rst [hunk 15]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+``cls_bpf`` programs, it attaches only a single program for a given hook in
+``direct-action`` mode. Typically, in the traditional tc scheme, there is a
+split between classifier and action modules, where the classifier has one
+or more actions attached to it that are triggered once the classifier has a
+match. In the modern world for using tc in the software data path this model
+does not scale well for complex packet processing. Given tc BPF programs
+attached to ``cls_bpf`` are fully self-contained, they effectively fuse the
+parsing and action process together into a single unit and therefore thanks
```
Comment (joestringer) [inline Documentation/bpf.rst]: `single unit. Thanks to..`

(run-on sentence)

File: Documentation/bpf.rst [hunk 16]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+to ``cls_bpf``'s ``direct-action`` mode will just return the tc action verdict
```
Comment (joestringer) [inline Documentation/bpf.rst]: If you change the above, you also need `s/mode/mode, it/`

File: Documentation/bpf.rst [hunk 17]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+and terminate the processing pipeline immediately which allows for implementing
```
Comment (joestringer) [inline Documentation/bpf.rst]: `immediately. This allows..`

File: Documentation/bpf.rst [hunk 18]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+scalable programmable packet processing in the networking data path. ``cls_bpf``
```
Comment (joestringer) [inline Documentation/bpf.rst]: Suggestion:
`networking data path` *by avoiding linear iteration of actions*.

File: Documentation/bpf.rst [hunk 19]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+is the only such "classifier" module in the tc layer capable of such fast-path.
```
Comment (joestringer) [inline Documentation/bpf.rst]: `such` *a* `fast-path`.

File: Documentation/bpf.rst [hunk 20]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+
+Like XDP BPF programs, tc BPF programs can be atomically updated at runtime
+via ``cls_bpf`` without interrupting any network traffic or having to restart
+services.
+
+Both the tc ingress but also egress hook where ``cls_bpf`` itself can be
```
Comment (joestringer) [inline Documentation/bpf.rst]: `s/but also/and the/`

File: Documentation/bpf.rst [hunk 21]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+attached to is managed by a pseudo qdisc called ``sch_clsact``. The latter
```
Comment (joestringer) [inline Documentation/bpf.rst]: `s/The latter/This`

File: Documentation/bpf.rst [hunk 22]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+is a drop-in replacement and proper superset of the ingress qdisc since it
+is able to manage both, ingress and egress tc hooks. For tc's egress hook
+in ``__dev_queue_xmit()`` it is important to stress that it is not executed
+under the kernel's qdisc root lock. Thus, both tc ingress and egress hooks
+are executed in a lockless manner in the fast-path. In either case, preemption
+is disabled and execution happens under RCU read side.
```
Comment (joestringer) [inline Documentation/bpf.rst:3472]: :champagne:

File: Documentation/bpf.rst [hunk 23]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+perform the heavy lifting packet classification through tc BPF outside of the
+qdisc root lock, setting the ``skb->mark`` or ``skb->priority`` from there such
+that ``sch_htb`` only requires a flat mapping without expensive packet
+classification under the root lock thus reducing contention.
+
+Offloaded tc BPF programs are supported for the case of ``sch_clsact`` in
+combination with ``cls_bpf`` where the prior loaded BPF program was JITed
+from a SmartNIC driver to be run natively on the NIC. Only ``cls_bpf``
+programs operating in ``direct-action`` mode are supported to be offloaded.
+``cls_bpf`` can also only be offloaded in the single program case and does
```
Comment (joestringer) [inline Documentation/bpf.rst]: `s/can also only be offloaded in the single program case and does not work for multiple programs,/only supports offloading a single program and cannot offload multiple programs./`

File: Documentation/bpf.rst [hunk 24]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+not work for multiple programs, and furthermore only the ingress hook is
```
Comment (joestringer) [inline Documentation/bpf.rst]: `, and ...`
=>
`Furthermore only the ingress hook supports offloading BPF programs.`

File: Documentation/bpf.rst [hunk 25]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+where the ``TC_ACT_UNSPEC`` from there continues with a next tc BPF program
+solely running in non-offloaded case. Last but not least, iii) ``TC_ACT_UNSPEC``
+is also used for the single program case to simply tell the kernel to continue
+with the ``skb`` without additional side-effects. ``TC_ACT_UNSPEC`` is very
+similar to the ``TC_ACT_OK`` action code in the sense that both pass the
+``skb`` onwards either to upper layers of the stack on ingress or down to
+the networking device driver for transmission on egress, respectively. The
+only difference to ``TC_ACT_OK`` is that ``TC_ACT_OK`` sets ``skb->tc_index``
+based on the classid the tc BPF program set. The latter is set out of the
+tc BPF program itself through ``skb->tc_classid`` from the BPF context.
```
Comment (joestringer) [inline Documentation/bpf.rst:3540]: Bullet-point formatting here for (i), (ii), (iii) would help readability.

File: Documentation/bpf.rst [hunk 26]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+  never been largely relevant. Additionally, ``act_bpf`` does not provide a tc
+  offloading interface either compared to ``cls_bpf``.
+
+..
+
+* **Question:** Is it recommended to use ``cls_bpf`` not in ``direct-action`` mode?
+* **Answer:** No. The answer is similar to the one above in that this is otherwise
+  unable to scale for more complex processing. tc BPF can already do everything needed
+  by itself in an efficient manner and thus there is no need for anything other than
+  ``direct-action`` mode.
```
Comment (joestringer) [inline Documentation/bpf.rst:3594]: Subquestion: What if you hit BPF instruction limits? Could you work around those by chaining programs via cls actions? (Might not be common and may/may not be something you actually want to include in the FAQ, I'm mostly just curious for myself).

File: Documentation/bpf.rst [hunk 27]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+
+  The forwarding and load-balancing use case is quite similar to XDP, although
+  slightly more targeted towards east-west container workloads rather than
+  north-south traffic (though both technologies can be used in either case).
+  Since XDP is only available on ingress side, tc BPF programs allow for
+  further use cases that apply in particular on egress, for example, container
+  based traffic can already be NATed and load-balanced on the egress side
+  through BPF out of the initial namespace such that this is done transparent
+  to the container itself. Egress traffic is due to the nature of the kernel's
+  networking stack already ``skb`` based anyway such that packet rewrites and
```
Comment (joestringer) [inline Documentation/bpf.rst]: Rearrange a bit, 

```
Egress traffic is already based on the ``sk_buff`` structure due to the nature of the kernel's networking stack, so packet rewrites...
```

File: Documentation/bpf.rst [hunk 28]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+  redirects are suitable out of tc BPF. Through utilizing the ``bpf_redirect()``
```
Comment (joestringer) [inline Documentation/bpf.rst]: `s/Through/By/`

File: Documentation/bpf.rst [hunk 29]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+  helper function, BPF can take over the forwarding logic to push the ``skb``
```
Comment (joestringer) [inline Documentation/bpf.rst]: `s/``skb``/packet/`?

File: Documentation/bpf.rst [hunk 30]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+  high-performance lockless per-CPU memory mapped perf ring buffer where the
+  BPF program is able to push custom data, the full or truncated packet
+  contents, or both up to a user space application. From the tc BPF program
+  this is realized through the ``bpf_skb_event_output()`` BPF helper function
+  which has the same function signature and semantics as ``bpf_xdp_event_output()``.
+  Given tc BPF programs can be attached to ingress and egress as opposed to
+  only ingress in XDP BPF case plus the two tc hooks are at the lowest layer
+  in the (generic) networking stack, this allows for bidirectional monitoring
+  of all network traffic from a particular node.
+
```
Comment (joestringer) [inline Documentation/bpf.rst]: Could be worth mentioning here that this harkens back to old-school BPF and tcpdump.

File: Documentation/bpf.rst [hunk 31]
```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
...
+ones:
+
+* The BPF input context is a ``sk_buff`` not a ``xdp_buff``. The former is
+  the network packet representation of the kernel's networking stack and
+  therefore used in many places. As such it has all the relevant ``sk_buff``
+  metadata available the BPF program can read and write into, whereas in
+  ``xdp_buff`` case it's limited due to the early stage processing where the
+  cost of allocating a ``sk_buff`` is avoided altogether.
+
+  Out of the tc BPF hooks, the BPF program can for instance read or write
```
Comment (joestringer) [inline Documentation/bpf.rst]: Perhaps,

`Out of the tc BPF hooks, the BPF program can...`
=>
`BPF programs attached to the tc BPF hook can...`

**Review by joestringer — COMMENTED** [review]
Mostly minor comments, thanks for writing this up!

File: Documentation/bpf.rst (hunk 2 above)
Comment (borkmann) [reply Documentation/bpf.rst]: No that is correct.

File: Documentation/bpf.rst (hunk 26 above)
Comment (borkmann) [reply Documentation/bpf.rst:3594]: Yes, this would work. You then need to transfer state either via skb->cb or per-cpu map with single slot.

File: Documentation/bpf.rst (hunk 6 above)
Comment (borkmann) [reply Documentation/bpf.rst]: I think it's probably okay as is, tc BPF programs meaning in a more general sense here.

File: Documentation/bpf.rst (hunk 3 above)
Comment (borkmann) [reply Documentation/bpf.rst]: Much better indeed, I've slightly modified it while integrating.

**Comment by borkmann** [comment]
test-docs-please

**Comment by borkmann** [comment]
test-docs-please

**Comment by borkmann** [comment]
test-docs-please

**Comment by borkmann** [comment]
test-docs-please

**Comment by borkmann** [comment]
test-docs-please

**Comment by borkmann** [comment]
Thanks @joestringer for the awesome reviews, very much appreciated! I've addressed them, and am merging this in a bit for getting this deployed asap to have full doc coverage from the ref guide. Discussed previously with @aanm that we're using `sort -ud` for the spellcheck list going forward which this is doing here.

File: Documentation/bpf.rst [hunk 32]
```diff
@@ -3354,7 +3354,376 @@ the toolchain section under the respective tools.
 tc (traffic control)
 --------------------
 
-TODO
+Aside from other program types such as XDP, BPF can also be used out of the
+kernel's tc (traffic control) layer in the networking data path. On a high-level
+there are three major differences when comparing BPF XDP programs to tc BPF
+ones:
```
Comment (raybejjani) [inline Documentation/bpf.rst:3360]: `tc BPF` should be `BPF tc` to follow the other one, perhaps?

File: Documentation/bpf.rst [hunk 33]
```diff
@@ -3354,7 +3354,376 @@ the toolchain section under the respective tools.
...
+  simply switch protocols by solely rewriting the packet data. This is due to
+  the stack processing the packet based on the metadata rather than having the
+  cost of accessing the packet contents each time. Thus, additional conversion
+  is required from BPF helper functions taking care that ``sk_buff`` internals
+  are properly converted as well. The ``xdp_buff`` case however does not
+  face such issues since it comes at such an early stage where the kernel
+  has not even allocated an ``sk_buff`` yet, thus packet rewrites of any
+  kind can be realized trivially. However, the ``xdp_buff`` case has the
+  disadvantage that ``sk_buff`` metadata is not available for mangling
+  at this time. The latter is overcome by passing custom metadata from
```
Comment (raybejjani) [inline Documentation/bpf.rst:3395]: `at this stage` since you used the word above and "at this time" sounds like you're saying: today this isn't supported but it might be later.

File: Documentation/bpf.rst [hunk 34]
```diff
@@ -3354,7 +3354,376 @@ the toolchain section under the respective tools.
...
+  XDP BPF to tc BPF, though. In this way, the limitations of each program
+  type can be overcome by operating complementary programs of both types
+  as the use case requires.
+
+..
+
+* Compared to XDP, tc BPF programs can be triggered out of ingress and also
+  egress points in the networking data path as opposed to ingress only in case
```
Comment (raybejjani) [inline Documentation/bpf.rst:3403]: `only in the case of XDP`

File: Documentation/bpf.rst [hunk 35]
```diff
@@ -3354,7 +3354,376 @@ the toolchain section under the respective tools.
...
+  of XDP.
+
+  The two hook points ``sch_handle_ingress()`` and ``sch_handle_egress()`` in
+  the kernel are triggered out of ``__netif_receive_skb_core()`` and
+  ``__dev_queue_xmit()``, respectively. The latter two are the main receive
+  and transmit functions in the data path that, taking XDP aside, are triggered
```
Comment (raybejjani) [inline Documentation/bpf.rst:3409]: "functions in the data path that, along with XDP, are triggered" although it's clear what your saying anyway.

File: Documentation/bpf.rst [hunk 36]
```diff
@@ -3354,7 +3354,376 @@ the toolchain section under the respective tools.
...
+
+  For virtual devices like veth devices XDP is unsuitable in this case since the
+  kernel operates solely on a ``skb`` here and generic XDP has a few limitations
+  where it does not operate with cloned ``skb``'s. The latter is heavily used
+  from the TCP/IP stack in order to hold data segments for retransmission where
+  the generic XDP hook would simply get bypassed instead. Moreover, generic XDP
+  needs to linearize the entire ``skb`` resulting in heavily degraded performance.
+  tc BPF on the other hand is more flexible as it specializes on the ``skb``
+  input context case and thus does not need to cope with the limitations from
+  generic XDP.
```
Comment (raybejjani) [inline Documentation/bpf.rst:3641]: Is the intended audience once that knows what an `skb` is? we use `sk_buff` everywhere else, it seems?

**Review by raybejjani — COMMENTED** [review]
Stylistic things that can be ignored :)

File: Documentation/bpf.rst (hunk 36 above)
Comment (borkmann) [reply Documentation/bpf.rst:3641]: Yes, they should know.

File: Documentation/bpf.rst (hunk 35 above)
Comment (borkmann) [reply Documentation/bpf.rst:3409]: `along with XDP` would not be fully correct. XDP doesn't exist on egress. And XDP ingress could drop packets where they are never seen on tc ingress. Meaning, if you take XDP aside for a moment, then every packet passes through tc ingress and egress hook.

File: Documentation/bpf.rst (hunk 32 above)
Comment (borkmann) [reply Documentation/bpf.rst:3360]: Should say XDP BPF

File: Documentation/bpf.rst (hunk 35 above)
Comment (raybejjani) [reply Documentation/bpf.rst:3409]: ah, ok, then say "setting XDP aside" or "not considering XDP"

Comment (borkmann) [reply Documentation/bpf.rst:3409]: Good point, done.
